core/app/models/spree/order_contents.rb

```diff
@@ -57,6 +71,20 @@ def approve(user: nil, name: nil)
 
     private
 
+    def should_recalculate_taxes?(old_address, new_address)
+      # Possible cases:
+      # [Old_address is a TaxLocation, New_address is a TaxLocation ]:
+      # NON-VAT order before Address state : we don't compute taxes.
+      # [Old_address is a TaxLocation, New_address is an Adress ]:
+      # Order who passed Address state : we compute taxes.
+      # [Old_address is an Adress, New_address is a TaxLocation ]:
+      # Should not happen, TaxLocation can be replaced by an Address but not the other way around.
+      # Order is a weird state and we do nothing because it should not happen.
+      return if old_address.is_a?(Spree::Tax::TaxLocation) || new_address.is_a?(Spree::Tax::TaxLocation)
+
+      old_address.try(:taxation_attributes) != new_address.try(:taxation_attributes)
```

Comment:
Comments look wrong, so I need to fix that but otherwise : 

I don't think TaxLocation needs to implement taxation_attributes because we're not supposed to compare TaxLocations for the following reasons : if old_address is a TaxLocation it means that that it's either a VAT order and it's handled elsewhere or it's a standard order who hasn't reached the Address state so the taxes will be handled by the checkout

The only case where this is suppose to apply is when you have two addresses, so this would means that you're either updating the customer details from the backend, or that a registered user is updating his default addresses so the taxes showing in the cart would be updated.

Does it make mores sense? 

Good point about using try! though, we can be extra sure that we're comparing Spree::Addresses because it will crash otherwise